src/app/ReadClient.cpp

```diff
@@ -914,48 +954,108 @@ CHIP_ERROR ReadClient::SendSubscribeRequestImpl(const ReadPrepareParams & aReadP
+    VerifyOrDie(_this != nullptr);
+
+    ChipLogError(DataManagement, "Failed to establish CASE for re-subscription with error '%" CHIP_ERROR_FORMAT "'", err.Format());
+
+    _this->Close(err);
+}
+
+void ReadClient::OnResubscribeTimerCallback(System::Layer * apSystemLayer, void * apAppState)
+{
+    ReadClient * const _this = static_cast<ReadClient *>(apAppState);
+    VerifyOrDie(_this != nullptr);
+
+    CHIP_ERROR err;
+
+    ChipLogProgress(DataManagement, "OnResubscribeTimerCallback: DoCASE = %d", _this->mDoCaseOnNextResub);
+    _this->mNumRetries++;
+
+    if (_this->mDoCaseOnNextResub)
     {
-        ChipLogProgress(DataManagement, "Resubscribe has been stopped");
-        return false;
+        auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
+        VerifyOrExit(caseSessionManager != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+        //
+        // We need to mark our session as defunct explicitly since the assessment of a liveness failure
+        // is usually triggered by the absence of any exchange activity that would have otherwise
+        // automatically marked the session as defunct on a response timeout.
+        //
+        // Doing so will ensure that the subsequent call to FindOrEstablishSession will not bind to
+        // an existing established session but rather trigger establishing a new one.
+        //
+        if (_this->mReadPrepareParams.mSessionHolder)
+        {
+            _this->mReadPrepareParams.mSessionHolder.Get().Value()->AsSecureSession()->MarkAsDefunct();
+        }
+
+        //
+        // TODO: Until #19259 is merged, we cannot actually just get by with the above logic since marking sessions
+        //       defunct has no effect on resident OperationalDeviceProxy instances that are already bound
+        //       to a now-defunct CASE session.
+        //
+        auto proxy = caseSessionManager->FindExistingSession(_this->mPeer);
+        if (proxy != nullptr)
+        {
+            proxy->Disconnect();
+        }
+
+        caseSessionManager->FindOrEstablishSession(_this->mPeer, &_this->mOnConnectedCallback,
+                                                   &_this->mOnConnectionFailureCallback);
+        return;
     }
-    CHIP_ERROR err = InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->StartTimer(
-        System::Clock::Milliseconds32(intervalMsec), OnResubscribeTimerCallback, this);
+
+    err = _this->SendSubscribeRequest(_this->mReadPrepareParams);
+
+exit:
     if (err != CHIP_NO_ERROR)
     {
-        ChipLogError(DataManagement, "Fail to resubscribe with error %" CHIP_ERROR_FORMAT, err.Format());
-        return false;
+        //
+        // Call Close (which should trigger re-subscription again) EXCEPT if we got here because we didn't have a valid fabric,
+        // or an invalid CASESessionManager pointer when mDoCaseOnNextResub was true.
```

Comment:
```suggestion
        // Call Close (which should trigger re-subscription again) EXCEPT if we got here because we didn't have a valid
        // CASESessionManager pointer when mDoCaseOnNextResub was true.
```